pkg/apis/storage/types.go

```diff
@@ -424,3 +436,76 @@ type CSINodeList struct {
 	// items is the list of CSINode
 	Items []CSINode
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIStorageCapacity stores the result of one CSI GetCapacity call for one
+// driver, one topology segment, and the parameters of one storage class.
+//
+// This is an alpha feature and only available when the CSIStorageCapacity feature is enabled.
+type CSIStorageCapacity struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning and just has to
+	// meet the usual requirements (length, characters, unique). To ensure that
+	// there are no conflicts with other CSI drivers on the cluster, the recommendation
+	// is to use csisc-<uuid>, a generated name, or a reverse-domain name which ends
+	// with the unique CSI driver name.
+	//
+	// Objects are not namespaced.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Spec contains the fixed properties of one capacity value.
+	Spec CSIStorageCapacitySpec
+
+	// Status contains the properties that can change over time.
+	Status CSIStorageCapacityStatus
+}
+
+// CSIStorageCapacitySpec contains the fixed properties of one capacity value.
+type CSIStorageCapacitySpec struct {
+	// The CSI driver that provides the storage.
+	// This must be the string returned by the CSI GetPluginName() call.
+	DriverName string
+
+	// NodeTopology defines which nodes have access to the storage for which
+	// capacity was reported. If not set, the storage is accessible from all
+	// nodes in the cluster.
+	//
+	// +optional
+	NodeTopology *api.NodeSelector
```

Comment:
By allowing `api.NodeSelector` here, we allow matching by fields (https://pkg.go.dev/k8s.io/api/core/v1?tab=doc#NodeSelectorTerm). But *which* fields?

I found that the set of supported fields is going to be defined by the implementation and how it calls `MatchNodeSelectorTerms`, similar to https://github.com/kubernetes/kubernetes/blob/dcd0755f84bf8c15fd21e8d697b924991f641ae4/pkg/scheduler/framework/plugins/helper/node_affinity.go#L72-L78

Supporting just node name as field doesn't give us any additional functionality because there is also a label for it. Are there other fields we might want to filter on? We need to document all of them here.

I'm tempted to just say that matching by fields is not supported. Then we can either use `api.NodeSelector` and validate that all `MatchFields` are nil, or create a custom type that doesn't have `MatchFields` at all.

Thoughts?

/hold

@msau42

Reply:
I agree, so let's document that fields are not supported and validated accordingly.

Reply:
Done, updated API pushed here.

Reply:
If this only matches labels, do we really need to reuse this type?  Why not a simple `metav1.LabelSelector`?  That simpler and more consistent and avoids my argument about cloning this type.

Reply:
I've switched to `metav1.LabelSelector`.